Implemented cached settings manager

Nyholm opened the pull request:
This PR address the issue in #8  The PR is currently missing documentation. 

Question: Should the `CachedManager` extend `SettingsManager`? The `CacheManager` could just have an instance of a `SettingsManager` object. What are the pros and cons?

Nyholm commented:
To answer my own question: We want loose coupling. The `CacheManager` should not extend the `SettingsManager`.

Nyholm commented:
There you go. I'm done with this PR.

Nyholm commented:
Fixed the merge conflicts.

Nyholm commented:
Travis gives a false negative. It's composer self-update that fails on HHVM. if you restart it you will see that everything is green.

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Manager/CachedManager.php:
```diff
@@ -0,0 +1,188 @@
+
+        $value = $this->settingsManager->all($owner);
+
+        //save
```
Same as above

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Manager/CachedManager.php:
```diff
@@ -0,0 +1,188 @@
+
+        $value = $this->settingsManager->get($name, $owner, $default);
+
+        //save
```
I would prefer not to use too obvious comments

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Tests/CachedManagerTest.php:
```diff
@@ -0,0 +1,197 @@
+
+use Dmishh\Bundle\SettingsBundle\Entity\SettingOwner;
+use Dmishh\Bundle\SettingsBundle\Manager\CachedManager;
+use Mockery as m;
```
What is the purpose of this alias?

Nyholm replied:
It's just easier to write `m::mock` than `\Mocker:mock`. No other purpose.

Nyholm commented:
I removed those comments. I think I needed them because it was too confusing with the functions get, fetch, store, set... I renamed the protected functions to make it more clear.

Nyholm commented:
Are you interested in merging this PR? Should I rebase?

dmishh commented:
Definitely interested.
Thanks for contributing, Tobias!

Nyholm commented:
I've now rebased this and made some adjustments to the new master. IE renamed SettingsOwner to SettingsOwnerInterface in the CacheManager.php.

dmishh commented on README.md:
```diff
@@ -354,6 +379,9 @@ Please, do not hesitate to [report bugs](https://github.com/dmishh/SettingsBundl
 
 #### Version 2.0.0-dev
 
+Make sure to read the [Upgrade.md](Upgrade.md) to successfully migrate your application.
```
`UPGRADE.md` (all capitals)

dmishh commented on README.md:
```diff
@@ -253,6 +255,29 @@ dmishh_settings:
+    cache_lifetime: 3600
+
+doctrine_cache:
+  aliases:
```
4 spaces indentation

dmishh commented on composer.json:
```diff
@@ -26,7 +26,7 @@
         "symfony/validator": "~2.1,!=2.5.0",
         "symfony/translation": "~2.1,!=2.5.0",
-        "doctrine/orm": "~2.2,>=2.2.3"
+        "doctrine/orm": "~2.2,>=2.2.3,<2.5"
```
I don't think that it's a good idea. We can just leave `2.4.x` installed in `composer.lock`, but without restrictions here

Nyholm replied:
Hm, I did like that because this is how Symfony does it. https://github.com/symfony/symfony-standard/blob/2.7/composer.json#L12

I improved the .travis.yml instead.. lets see if it works.

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Tests/CachedManagerTest.php:
```diff
@@ -0,0 +1,195 @@
+
+use Dmishh\Bundle\SettingsBundle\Entity\SettingsOwnerInterface;
+use Dmishh\Bundle\SettingsBundle\Manager\CachedManager;
+use Mockery as m;
```
I suggest using full class name `\Mockery`

Nyholm replied:
Sure!

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Tests/CachedManagerTest.php:
```diff
@@ -0,0 +1,195 @@
+        $value = 'foobar';
+        $defaultValue = 'default';
+
+        $sm = m::mock('Dmishh\Bundle\SettingsBundle\Manager\SettingsManager');
```
Let's use extended name for this var: `$settingsManager`

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Tests/CachedManagerTest.php:
```diff
@@ -0,0 +1,195 @@
+        $sm = m::mock('Dmishh\Bundle\SettingsBundle\Manager\SettingsManager');
+        $sm->shouldReceive('get')->once()->with($name, $owner, $defaultValue)->andReturn($value);
+
+        $manager = $this->getMockBuilder('Dmishh\Bundle\SettingsBundle\Manager\CachedManager')
```
Let's use extended name for this var: `$cachedSettingsManager`

dmishh replied:
I think, code will become more clear and easier to read, won't it?

Nyholm replied:
Agreed

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Manager/CachedManager.php:
```diff
@@ -0,0 +1,185 @@
+ */
+class CachedManager implements SettingsManagerInterface
+{
+    const PREFIX = 'dmishh_o{%s}_v{%s}';
```
Maybe change to `dmishh_settings_o{%s}_k{%}` ?
o — owner, k — key

Nyholm replied:
Sure, This is actually the second prefix. The proper one is configured when you define your cache service. So, `{%s}_{%s}` would be enough. But since we are lazy and don't want to define too many cache services I've added the `dmishh_`

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Manager/CachedManager.php:
```diff
@@ -0,0 +1,185 @@
+/**
+ * @author Tobias Nyholm
+ */
+class CachedManager implements SettingsManagerInterface
```
I suggest renaming this to `CachedSettingsManager`. What do you think, Tobias?

Nyholm replied:
Agreed

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Tests/CachedManagerTest.php:
```diff
@@ -0,0 +1,195 @@
+        $manager->clear($name, $owner);
+    }
+
+    public function testGetCacheKey()
```
I think you can skip testing this protected method, because it covered by other tests and method itself is quite simple. Also, `CachedDummy` can be removed

Nyholm replied:
I wrote this test because I found a bug that the other test's did not cover. It makes sure that it always returns a string no matter the input.

dmishh replied:
Ok, but then let's remove `dummy` and use reflection. But most important is to write the purpose of this test in a comment.

Nyholm commented:
I've done changes according to your comments, is this ready to merge?

Nyholm commented:
Ping @dmishh

dmishh commented:
Hi @Nyholm Good job! Sorry for late answers, I'm quite busy now.

Nyholm commented:
Sure, We are almost done with this PR. The other two (#29 and #30) are "no brainers". I'm planning to support PSR-4 and also fix #24 and possible #31 if you want that feature.

dmishh commented on src/Dmishh/Bundle/SettingsBundle/Tests/CachedSettingsManagerTest.php:
```diff
@@ -0,0 +1,188 @@
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        // make sure we do always return a string, no matter input
```
Nice, but let's move this to the method's description, so we will be able to figure out instantly what the test is about

Nyholm commented:
There you go, updated according to your comments.

dmishh commented:
:+1:
